pkg/streamingpromql/operators/aggregations/topkbottomk/range_query.go

```diff
@@ -0,0 +1,505 @@
+	nextSeries, err := t.Inner.NextSeries(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(nextSeries.Histograms) > 0 {
+		t.emitIgnoredHistogramsAnnotation()
+	}
+
+	// topk() and bottomk() ignore histograms, so return the HPoint slice to the pool now.
+	types.HPointSlicePool.Put(nextSeries.Histograms, t.MemoryConsumptionTracker)
+
+	g := t.remainingInnerSeriesToGroup[0]
+	t.remainingInnerSeriesToGroup = t.remainingInnerSeriesToGroup[1:]
+	if err := t.accumulateIntoGroup(nextSeries, g); err != nil {
+		return err
+	}
+
+	types.FPointSlicePool.Put(nextSeries.Floats, t.MemoryConsumptionTracker)
+
+	return nil
+}
+
+func (t *RangeQuery) emitIgnoredHistogramsAnnotation() {
+	if t.haveEmittedIgnoredHistogramsAnnotation {
+		return
+	}
+
+	if t.IsTopK {
+		t.annotations.Add(annotations.NewHistogramIgnoredInAggregationInfo("topk", t.expressionPosition))
+	} else {
+		t.annotations.Add(annotations.NewHistogramIgnoredInAggregationInfo("bottomk", t.expressionPosition))
+	}
+
+	t.haveEmittedIgnoredHistogramsAnnotation = true
+}
+
+func (t *RangeQuery) accumulateIntoGroup(data types.InstantVectorSeriesData, g *rangeQueryGroup) error {
+	groupSeriesIndex := g.seriesRead()
+
+	if g.seriesForTimestamps == nil {
+		var err error
+		g.seriesForTimestamps, err = intSliceSlicePool.Get(t.TimeRange.StepCount, t.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.seriesForTimestamps = g.seriesForTimestamps[:t.TimeRange.StepCount]
+	}
+
+	if g.series == nil {
+		var err error
+		g.series, err = rangeQuerySeriesSlicePool.Get(g.totalSeries, t.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+	}
+
+	g.series = g.series[:groupSeriesIndex+1]
+	thisSeries := &g.series[groupSeriesIndex]
+
+	for _, p := range data.Floats {
+		timestampIndex := t.TimeRange.PointIndex(p.T)
+		limit := t.k[timestampIndex]
+
+		if limit == 0 {
+			continue
+		}
+
+		if g.seriesForTimestamps[timestampIndex] == nil {
+			// This is the first time we've seen a point for this timestamp, create the list of source series.
+			maximumPossibleSeries := min(limit, int64(g.totalSeries))
+
+			var err error
+			g.seriesForTimestamps[timestampIndex], err = types.IntSlicePool.Get(int(maximumPossibleSeries), t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+		}
+
+		populateSeriesSlices := func() error {
+			lastPointIndex := t.TimeRange.PointIndex(data.Floats[len(data.Floats)-1].T)
+			sliceLength := int(lastPointIndex + 1)
+
+			var err error
+			thisSeries.shouldReturnPoint, err = types.BoolSlicePool.Get(sliceLength, t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+
+			thisSeries.shouldReturnPoint = thisSeries.shouldReturnPoint[:sliceLength]
+
+			thisSeries.values, err = types.Float64SlicePool.Get(sliceLength, t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+
+			thisSeries.values = thisSeries.values[:sliceLength]
+
+			return nil
+		}
+
+		recordUsedPoint := func() error {
```

Comment:
ditto the positioning re above

Reply:
This method uses `timestampIndex`, which is declared inside the loop. 

We could move the method out and pass `timestampIndex` to the method, but I don't have a strong preference either way.

Reply:
My gut feel is passing timestampIndex is better, but I don't know how much difference it'd actually make. My preference is to move it, but I'm not blocked on it.